### ISSUE TITLE
attempt to index field 'lsp' (a nil value)

### ISSUE THREAD

**Issue by Livingnight** [LunarVim/LunarVim#4031]
### Problem description

When using tailwindcss I get an lsp error on all files in a project with tailwindcss LSP server attached

### LunarVim version

master-1e1e7bb4

### Neovim version (>= 0.8.0)

NVIM v0.8.3 & release-0.9

### Operating system/version

Windows 11 WSL2 Ubuntu 20.04

### Affected language servers

tailwindcss

### Steps to reproduce

setup a basic lunarvim config with tailwindcss LSP server installed and enter a file that uses tailwind to see the error

### Actual behavior

An error shows on all files with tailwindcss lsp server attached to the buffer. Dynamic registration causes cascaded of weirdness with all attached servers.

### Expected behavior

No errors with tailwind server attached. dynamic registration handled gracefully.

### support info

```console
 Language client log: /home/livingnight/.local/state/nvim/lsp.log
 Detected filetype:   typescriptreact
 
 4 client(s) attached to this buffer: 
 
 Client: tsserver (id: 1, bufnr: [11])
 	filetypes:       javascript, javascriptreact, javascript.jsx, typescript, typescriptreact, typescript.tsx
 	autostart:       true
 	root directory:  /home/livingnight/dev/projects/open-table-clone
 	cmd:             typescript-language-server --stdio
 
 Client: tailwindcss (id: 2, bufnr: [11])
 	filetypes:       html, vue, typescriptreact, javascriptreact
 	autostart:       true
 	root directory:  /home/livingnight/dev/projects/open-table-clone
 	cmd:             tailwindcss-language-server --stdio
 
 Client: emmet_ls (id: 3, bufnr: [11])
 	filetypes:       astro, html, javascriptreact, typescriptreact
 	autostart:       true
 	root directory:  /home/livingnight/dev/projects/open-table-clone
 	cmd:             emmet-ls --stdio
 
 Client: null-ls (id: 4, bufnr: [11, 52])
 	filetypes:       go, python, typescriptreact, javascriptreact, javascript, scss, typescript, css, zsh, sh, bash, java, lua, toml
 	autostart:       false
 	root directory:  /home/livingnight/dev/projects/open-table-clone
 	cmd:             <function>
 
 2 active client(s) not attached to this buffer: 
 
 Client: lua_ls (id: 5, bufnr: [52])
 	filetypes:       lua
 	autostart:       true
 	root directory:  Running in single file mode.
 	cmd:             lua-language-server
 
 Client: jsonls (id: 6, bufnr: [63])
 	filetypes:       json, jsonc
 	autostart:       true
 	root directory:  Running in single file mode.
 	cmd:             vscode-json-language-server --stdio
 
```
```console

    __                          _    ___         
   / /   __  ______  ____ _____| |  / (_)___ ___ 
  / /   / / / / __ \/ __ `/ ___/ | / / / __ `__ \
 / /___/ /_/ / / / / /_/ / /   | |/ / / / / / / /
/_____/\__,_/_/ /_/\__,_/_/    |___/_/_/ /_/ /_/ 


Buffer info
* filetype:                typescriptreact
* bufnr:                   13
* treesitter status:       inactive

Active client(s)
* name:                      tsserver
* id:                        1
* supported filetype(s):     [javascript, javascriptreact, javascript.jsx, typescript, typescriptreact, typescript.tsx]
* attached buffers:          13
* root_dir pattern:          13
* capabilities:              workspaceSymbolProvider         | inlayHintProvider      | documentSymbolProvider
                             referencesProvider              | foldingRangeProvider   | typeDefinitionProvider
                             documentRangeFormattingProvider | hoverProvider          | documentHighlightProvider
                             documentFormattingProvider      | definitionProvider
                             selectionRangeProvider          | implementationProvider
* name:                      tailwindcss
* id:                        2
* supported filetype(s):     [aspnetcorerazor, astro, astro-markdown, blade, clojure, django-html, htmldjango, edge, eelixir, elixir, ejs, erb, eruby, gohtml, haml, handlebars, hbs, html, html-eex, heex, jade, leaf, liquid, markdown, mdx, mustache, njk, nunjucks, php, razor, slim, twig, css, less, postcss, sass, scss, stylus, sugarss, javascript, javascriptreact, reason, rescript, typescript, typescriptreact, vue, svelte]
* attached buffers:          13
* root_dir pattern:          13
* capabilities:              hoverProvider | codeActionProvider | colorProvider

Automatic LSP info
* Skipped servers: [angularls, ansiblels, antlersls, csharp_ls, cssmodules_ls, denols, docker_compose_language_service, ember, emmet_ls, eslint, glint, golangci_lint_ls, gradle_ls, graphql, jedi_language_server, ltex, neocmake, phpactor, psalm, pylsp, pyre, quick_lint_js, reason_ls, rnix, rome, ruby_ls, ruff_lsp, solang, solc, sorbet, sourcery, spectral, sqlls, sqls, stylelint_lsp, svlangserver, tflint, unocss, verible, vtsls, vuels, rust_analyzer, gopls, tsserver, bashls, jdtls]

Formatters info
* Active: prettierd  
* Supported: [deno_fmt, dprint, eslint, eslint_d, prettier, prettier_d_slim, prettier_eslint, prettierd, rome, rustywind, standardts]

Linters info
* Active: eslint_d  
* Supported: [eslint, eslint_d, semgrep, tsc, xo]

Code actions info
* Active: 
```

### logs

```console
[START][2023-04-12 10:44:55] LSP logging initiated
[WARN][2023-04-12 10:44:55] ...lsp/handlers.lua:487	'Using Typescript version (workspace) 4.9.5 from path "/home/livingnight/dev/volta/volta_agency_site/node_modules/typescript/lib/tsserver.js"'
[WARN][2023-04-12 10:44:55] ...lsp/handlers.lua:113	"The language server emmet_ls triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
[WARN][2023-04-12 10:44:55] ...lsp/handlers.lua:113	"The language server emmet_ls triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
[WARN][2023-04-12 10:45:22] ...lsp/handlers.lua:487	'Using Typescript version (bundled) 5.0.2 from path "/home/livingnight/.local/share/nvim/mason/packages/typescript-language-server/node_modules/typescript/lib/tsserver.js"'
[WARN][2023-04-12 10:45:23] ...lsp/handlers.lua:113	"The language server tailwindcss triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
[WARN][2023-04-12 10:46:30] ...lsp/handlers.lua:113	"The language server tailwindcss triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
[WARN][2023-04-12 10:46:30] .../lua/vim/lsp.lua:1065	"server_request: no handler found for"	"client/unregisterCapability"
[WARN][2023-04-12 10:47:49] .../lua/vim/lsp.lua:1065	"server_request: no handler found for"	"client/unregisterCapability"
[WARN][2023-04-12 10:47:50] ...lsp/handlers.lua:113	"The language server tailwindcss triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
[WARN][2023-04-12 11:04:27] ...lsp/handlers.lua:487	'Failed to modify settings\n\nPlease modify following settings manually:\n* `Lua.workspace.library`: add element `"${3rd}/luv/library"` ;\n'
[WARN][2023-04-12 11:07:31] .../lua/vim/lsp.lua:1065	"server_request: no handler found for"	"client/unregisterCapability"
[START][2023-04-12 11:07:39] LSP logging initiated
[WARN][2023-04-12 11:07:39] ...lsp/handlers.lua:487	'Using Typescript version (workspace) 5.0.4 from path "/home/livingnight/dev/projects/open-table-clone/node_modules/typescript/lib/tsserver.js"'
[WARN][2023-04-12 11:07:39] ...lsp/handlers.lua:113	"The language server emmet_ls triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
[WARN][2023-04-12 11:07:39] ...lsp/handlers.lua:113	"The language server emmet_ls triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
[WARN][2023-04-12 11:07:40] ...lsp/handlers.lua:113	"The language server tailwindcss triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
[ERROR][2023-04-12 11:10:36] ...lsp/handlers.lua:485	"Unhandled exception: Cannot read properties of undefined (reading 'colorDecorators')\nTypeError: Cannot read properties of undefined (reading 'colorDecorators')\n    at Object.<anonymous> (/home/livingnight/.local/share/nvim/mason/packages/tailwindcss-language-server/node_modules/@tailwindcss/language-server/bin/tailwindcss-language-server:2963:705)\n    at Generator.next (<anonymous>)\n    at /home/livingnight/.local/share/nvim/mason/packages/tailwindcss-language-server/node_modules/@tailwindcss/language-server/bin/tailwindcss-language-server:2:1219\n    at new Promise (<anonymous>)\n    at Tn (/home/livingnight/.local/share/nvim/mason/packages/tailwindcss-language-server/node_modules/@tailwindcss/language-server/bin/tailwindcss-language-server:2:1039)\n    at Object.onUpdateSettings (/home/livingnight/.local/share/nvim/mason/packages/tailwindcss-language-server/node_modules/@tailwindcss/language-server/bin/tailwindcss-language-server:2963:656)\n    at wce.<anonymous> (/home/livingnight/.local/share/nvim/mason/packages/tailwindcss-language-server/node_modules/@tailwindcss/language-server/bin/tailwindcss-language-server:2963:13278)\n    at Generator.next (<anonymous>)\n    at i (/home/livingnight/.local/share/nvim/mason/packages/tailwindcss-language-server/node_modules/@tailwindcss/language-server/bin/tailwindcss-language-server:2:1077)"
[WARN][2023-04-12 11:10:54] .../lua/vim/lsp.lua:1065	"server_request: no handler found for"	"client/unregisterCapability"
[START][2023-04-12 11:10:59] LSP logging initiated
[WARN][2023-04-12 11:10:59] ...lsp/handlers.lua:113	"The language server emmet_ls triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
[WARN][2023-04-12 11:10:59] ...lsp/handlers.lua:113	"The language server emmet_ls triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
[WARN][2023-04-12 11:10:59] ...lsp/handlers.lua:487	'Using Typescript version (workspace) 5.0.4 from path "/home/livingnight/dev/projects/open-table-clone/node_modules/typescript/lib/tsserver.js"'
[WARN][2023-04-12 11:11:03] ...lsp/handlers.lua:113	"The language server tailwindcss triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
[WARN][2023-04-12 11:15:06] ...lsp/handlers.lua:487	'Failed to modify settings\n\nPlease modify following settings manually:\n* `Lua.workspace.library`: add element `"${3rd}/luv/library"` ;\n'
[ERROR][2023-04-12 11:15:41] ...lsp/handlers.lua:485	"Unhandled exception: Cannot read properties of undefined (reading 'colorDecorators')\nTypeError: Cannot read properties of undefined (reading 'colorDecorators')\n    at Object.<anonymous> (/home/livingnight/.local/share/nvim/mason/packages/tailwindcss-language-server/node_modules/@tailwindcss/language-server/bin/tailwindcss-language-server:2963:705)\n    at Generator.next (<anonymous>)\n    at /home/livingnight/.local/share/nvim/mason/packages/tailwindcss-language-server/node_modules/@tailwindcss/language-server/bin/tailwindcss-language-server:2:1219\n    at new Promise (<anonymous>)\n    at Tn (/home/livingnight/.local/share/nvim/mason/packages/tailwindcss-language-server/node_modules/@tailwindcss/language-server/bin/tailwindcss-language-server:2:1039)\n    at Object.onUpdateSettings (/home/livingnight/.local/share/nvim/mason/packages/tailwindcss-language-server/node_modules/@tailwindcss/language-server/bin/tailwindcss-language-server:2963:656)\n    at wce.<anonymous> (/home/livingnight/.local/share/nvim/mason/packages/tailwindcss-language-server/node_modules/@tailwindcss/language-server/bin/tailwindcss-language-server:2963:13278)\n    at Generator.next (<anonymous>)\n    at i (/home/livingnight/.local/share/nvim/mason/packages/tailwindcss-language-server/node_modules/@tailwindcss/language-server/bin/tailwindcss-language-server:2:1077)"
```

### Screenshots

![Screenshot 2023-04-11 110515](https://user-images.githubusercontent.com/26702985/231505022-433081bd-5852-4b55-b276-06524d7df139.png)
![Screenshot 2023-04-11 112315](https://user-images.githubusercontent.com/26702985/231505024-03408f46-2e62-471b-ac9f-9b47c09dd74c.png)

**Comment by LostNeophyte** [comment]
does it happen with an empty config.lua?

**Comment by dastein1** [comment]
I just ran into a similar issue: switched to some older project and eslint (configured via null_ls) and got the same error.  
IMO, the issue is that the formatter tries to find a _code_ in a somewhat unsafe way (lua/lvim/lsp/config.lua#78):

      format = function(d)
        local code = d.code or (d.user_data and d.user_data.lsp.code)
        if code then
          return string.format("%s [%s]", d.message, code):gsub("1. ", "")
        end
        return d.message
      end,
 
In my case eslint fails early without an error code which makes `d.user_data.lsp` be nil.
Shall I send a PR that makes this check safer like so?

        local code = d.code or (d.user_data and d.user_data.lsp and d.user_data.lsp.code)

**Comment by cpea2506** [comment]
> I just ran into a similar issue: switched to some older project and eslint (configured via null_ls) and got the same error. IMO, the issue is that the formatter tries to find a _code_ in a somewhat unsafe way (lua/lvim/lsp/config.lua#78):
> 
> ```
>   format = function(d)
>     local code = d.code or (d.user_data and d.user_data.lsp.code)
>     if code then
>       return string.format("%s [%s]", d.message, code):gsub("1. ", "")
>     end
>     return d.message
>   end,
> ```
> 
> In my case eslint fails early without an error code which makes `d.user_data.lsp` be nil. Shall I send a PR that makes this check safer like so?
> 
> ```
>     local code = d.code or (d.user_data and d.user_data.lsp and d.user_data.lsp.code)
> ```

This was removed sinc #4070

**Comment by dastein1** [comment]
Oh, did not realize that. Good to know, unfortunately this fix is neither on the 1.2 or 1.3 branch.  
For people struggling with that, you can override this in your _config.lua_ via:

    lvim.lsp.diagnostics.float.format = nil
